Fix string/tuple/no auth on AsyncHttpConnection class

dannosaur opened the pull request:
### Description
Fixes an error and confusing code surrounding authentication on the AsyncHttpConnection class. aiohttp wants a `BasicAuth` instance instead of a tuple/colon-separated-string as requests does, so this accommodates what it wants also. The code in the `__init__` was lifted from the `RequestsHttpConnection` class as that's already a sensible implementation, but the way the resulting auth needs to be passed with the async session is slightly different.

### Issues Resolved
Closes #283

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on opensearchpy/connection/http_async.py:
```diff
@@ -65,10 +65,13 @@ def __init__(
                 http_auth = tuple(http_auth.split(":", 1))
 
+            if isinstance(http_auth, tuple):
+                http_auth = aiohttp.BasicAuth(login=http_auth[0], password=http_auth[1])
```
I think we are being too clever here. 

- We turn a tuple into a ... tuple. Seems unnecessary.
- We turn a list into a tuple.
- We turn a `string_types` into a tuple.
- Finally, we turn anything into `aiohttp.BasicAuth`.

Let's not be too clever, skip the intermediate `tuple` state and just make `aiohttp.BasicAuth` for any of the types?

```python
if isinstance(http_auth, tuple):
   http_auth = aiohttp.BasicAuth(login=http_auth[0], password=http_auth[1])
elif isinstance(http_auth, list)
   http_auth = aiohttp.BasicAuth(login=http_auth[0], password=http_auth[1])
elif isinstance(http_auth, string_types):
  http_auth = tuple(http_auth.split(":", 1))
```

dannosaur replied:
I tend to prefer DRY approaches, but that works too. I'll change it.

dblock replied:
Well, you choose what to DRY, `tuple` or `BasicAuth` :)

dannosaur replied:
Fair lol

dblock commented on opensearchpy/connection/http_async.py:
```diff
@@ -190,17 +193,23 @@ async def perform_request(
-        }
+        auth = None
+        if self._http_auth:
+            if isinstance(self._http_auth, tuple):
```
This shouldn't be possible anymore because `__init__` turned everything into `aiohttp.BasicAuth`. Is that callable, too?

dannosaur replied:
It _can_ be a callable if it's an instance of the AWS auth class, which doesn't belong in the `auth` parameter to the session, so it's covering that base.

This would simplify down to

```
auth = self._http_auth if isinstance(self._http_auth, aiohttp.BasicAuth) else None
```

We still need to maintain `auth=None` to the session if the AWS auth class was passed in.

dblock reviewed (changes requested):
This needs tests for the input scenarios between tuple, strings, etc. Also is this the same problem not in async?

dannosaur commented:
> This needs tests for the input scenarios between tuple, strings, etc. Also is this the same problem not in async?

Sync mode works just fine. When I was switching my project over to async and it refused to authenticate I went digging.

dannosaur commented:
I'll push changes and tests for this tomorrow.

dblock commented:
> I'll push changes and tests for this tomorrow.

Thank you for helping out @dannosaur! No rush.

dannosaur commented:
Updated. One question. How do I omit my email from this required DCO? I don't want that out in the public domain.

dannosaur commented:
> How do I omit my email from this required DCO? I don't want that out in the public domain.

Nevermind. Found the Github stuff for hiding email addresses and rebased/did the DCO once again.

New to this :)

dblock commented:
> New to this :)

🤔 🦖 ;)

dblock commented on opensearchpy/connection/http_async.py:
```diff
@@ -190,17 +191,20 @@ async def perform_request(
+            req_headers = {
+                **req_headers,
+                **self._http_auth(method, url, query_string, body),
+            }
```
I believe `auth` is already `None` (if not it should be), isn't it, and these are mutually exclusive?

```
if isinstance(self._http_auth, aiohttp.BasicAuth):
   auth = self._http_auth
elif callable(self._http_auth):
   req_headers = {
     **req_headers,
     **self._http_auth(method, url, query_string, body),
   }
```

Not a must have, just thinking about making it more readable.

dannosaur replied:
`auth` is used in `self.session.request` a bit further down. Doing it this way will, when a callable is used for auth, raise a `NameError` because `auth` isn't defined. `auth=None` is the default to `session.request()` yes, but the only other alternative is to make 2 branches to call `self.session.request` with the different iterations, which I think is a lot messier. Doing it this way allows a single call to `session.request(auth=auth)`, where `auth` is `None` (the default), or an instance of `aiohttp.BasicAuth`.

dblock commented on test_opensearchpy/test_connection.py:
```diff
@@ -502,6 +504,46 @@ def urlopen_raise(*_, **__):
         assert str(e.value) == "Wasn't modified!"
 
 
+class TestAsyncHttpConnection(TestCase):
```
Let's pull this into a separate file? Either make an `async` folder with a bunch of tests or `test_async_http_connection.py`?

dannosaur replied:
Ok. Sure. I wasn't sure where to put it since it looked like this class didn't have a single test to its name :)

dblock reviewed (changes requested):
Add to CHANGELOG, and consider the other suggestions? LGTM otherwise.

dannosaur commented on test_opensearchpy/test_async_http_connection.py:
I noticed in coverage that the actual calls in `perform_request` hadn't been covered despite tests existing. Turns out async testing is a little more involved than what I'm used to. LMK if you want this done a different way. It seems the "official" way (as far as I can tell) on mocking out aiohttp.ClientSession is with a custom context manager mock class, coming from `asynctest` as opposed to the regular `mock` package.

dblock replied:
I really don't know more than you about this, inherited this code :( Do what you think is the right way.

dannosaur replied:
In the interest of making future tests not have to import this context manager from this specific test file, I've moved it to the utils package.

dblock commented:
Linter is complaining :( https://github.com/opensearch-project/opensearch-py/actions/runs/5427802648/jobs/9871401633?pr=424

I'm ok to merge it as is, lmk when it's ready and tests are passing.

dblock commented:
This looks good! Thanks for hanging in here. Linter/CI are still failing, will you please fix? 

I changed an option on this repo and hopefully CI will auto-run on this PR for you when you push updates.

dblock commented:
> Also, on the Python 3.7 front, it appears the IsolatedAsyncioTestCase class was added to unittest in Python 3.8. Is it acceptable to disable those tests on Python <3.8, or would you prefer that I explore options to have them run in all supported Python version?

Ideally all tests should run in all versions unless it's not possible (e.g. uses a feature of Python 3.8+). In that case I would accept an open issue and disabling those tests for <3.8 with a proper explanation.

dannosaur commented:
> Ideally all tests should run in all versions unless it's not possible (e.g. uses a feature of Python 3.8+). In that case I would accept an open issue and disabling those tests for <3.8 with a proper explanation.

Yeah, OK. I solved the 3.7 issue, but now 3.5 is complaining. I'll open a separate issue to talk about supported Python version before going any further with this.

dannosaur commented:
Woohoo!

@dblock I did open the issue about the older versions of Python, however I spotted that every other async-related test a) lives in the `test_async` directory, and b) is skipped for Python <3.6. So I moved the file, and it appears that the entire suite of checks is happy across the board. I simplified it somewhat as well, not relying on external libraries and pretty much copied the same format as the other tests were doing.

Let me know if you need anything else.

dblock commented:
Great job 👏
